src/Storage/Notmuch.hs

```diff
@@ -49,22 +51,22 @@ setNotmuchMailTags dbpath m =
   case mailTagsToNotmuchTags m of
     Nothing -> pure $ Left "Tags are corrupt"
     Just nmtags ->
-      bracket (runExceptT (databaseOpen dbpath) >>= either (error . show) pure)
-              (void . runExceptT . databaseDestroy)
-              (\db -> tagsToMessage nmtags m db)
+      bracket (runExceptT (databaseOpen dbpath))
+              (runExceptT . traverse_ databaseDestroy)
+              (runExceptT . (either (throwError . show) pure >=> tagsToMessage nmtags m))
 
 tagsToMessage
-  :: [Tag] -> NotmuchMail -> Database RW -> IO (Either String NotmuchMail)
+  :: (MonadError String m, MonadIO m)
+  => [Tag] -> NotmuchMail -> Database RW -> m NotmuchMail
 tagsToMessage xs m db =
-  runExceptT
-    (findMessage db
-                 (view mailId m)) >>=
-  \case
-    Left e -> pure $ Left (show e)
-    Right Nothing -> pure $ Left "boop"
-    Right (Just msg) -> do
-      _ <- runExceptT (messageSetTags xs msg)
-      pure $ Right m
+  -- This is nasty.  We have to run the transformer so we can change the
+  -- error type.  Instead, we should update hs-notmuch to abstract over
+  -- an "AsNotmuchError" typeclass, then create our own error sum type
+  -- with an instance of it.
+  runExceptT (
+    findMessage db (view mailId m)
+    >>= traverse (\msg -> messageSetTags xs msg $> m)
+  ) >>= either (throwError . show) (maybe (throwError "boop") pure)
```

Comment:
Heh hahhaha... the "boop" leaked into the outside world. Should have actually been something useful like: Notmuch was not able to find Message with id: 932932329 . Is your notmuch setup working?

Reply:
Right.  I'll tackle that refactor later, too.